pkg/api/validate/dynamic/dynamic.go

```diff
@@ -617,34 +623,92 @@ func (dv *dynamic) ValidateSubnets(ctx context.Context, oc *api.OpenShiftCluster
+	var attached int
+	for _, subnet := range subnetByID {
+		if subnetHasNSGAttached(subnet) {
+			attached++
+		}
+	}
+
+	// all subnets have an attached NSG
+	if attached == len(subnetByID) {
+		dv.log.Info("all subnets are attached, BYO NSG")
+		return api.PreconfiguredNSGEnabled, nil // correct setup by customer
+	}
+
+	// no subnets have attached NSG, fallback
+	if attached == 0 {
+		dv.log.Info("no subnets are attached, no longer BYO NSG. Fall back to using cluster NSG.")
+		return api.PreconfiguredNSGDisabled, nil
+	}
+
+	// some subnets have NSGs attached, error out
+	return api.PreconfiguredNSGDisabled,
+		&api.CloudError{
+			StatusCode: http.StatusBadRequest,
+			CloudErrorBody: &api.CloudErrorBody{
+				Code:    api.CloudErrorCodeInvalidLinkedVNet,
+				Message: errMsgNSGNotProperlyAttached,
+			},
+		}
+}
+
+func (dv *dynamic) ValidateSubnets(ctx context.Context, oc *api.OpenShiftCluster, subnets []Subnet) error {
+	subnetByID, err := dv.createSubnetMapByID(ctx, subnets)
+	if err != nil {
+		return err
+	}
+
+	if oc.Properties.ProvisioningState == api.ProvisioningStateCreating {
+		if oc.Properties.NetworkProfile.PreconfiguredNSG == api.PreconfiguredNSGEnabled {
+			oc.Properties.NetworkProfile.PreconfiguredNSG, err = dv.checkByoNSG(subnetByID)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// we're parsing through the subnets slice, not the map because we'll return consistent error messages on creation
+	for _, s := range subnets {
+		ss := subnetByID[s.ID]
+
+		switch oc.Properties.ProvisioningState {
+		case api.ProvisioningStateCreating:
+			if subnetHasNSGAttached(ss) && oc.Properties.NetworkProfile.PreconfiguredNSG != api.PreconfiguredNSGEnabled {
+				expectedNsgID, err := getRecordedNsgID(oc, s.ID)
 				if err != nil {
 					return err
 				}
-
-				if !strings.EqualFold(*ss.SubnetPropertiesFormat.NetworkSecurityGroup.ID, expectedNsgID) {
-					return api.NewCloudError(
-						http.StatusBadRequest,
-						api.CloudErrorCodeInvalidLinkedVNet,
-						s.Path,
-						"The provided subnet '%s' is invalid: must not have a network security group attached.",
-						s.ID,
-					)
+				if !isTheSameNSG(*ss.SubnetPropertiesFormat.NetworkSecurityGroup.ID, expectedNsgID) {
+					return newSubnetValidationError(s.Path, errMsgNSGAttached, s.ID)
 				}
 			}
-		} else {
-			nsgID, err := subnet.NetworkSecurityGroupID(oc, *ss.ID)
+		default:
+			nsgID, err := getRecordedNsgID(oc, *ss.ID)
 			if err != nil {
 				return err
 			}
 
-			if ss.SubnetPropertiesFormat == nil ||
-				ss.SubnetPropertiesFormat.NetworkSecurityGroup == nil ||
-				!strings.EqualFold(*ss.SubnetPropertiesFormat.NetworkSecurityGroup.ID, nsgID) {
-				return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidLinkedVNet, s.Path, "The provided subnet '%s' is invalid: must have network security group '%s' attached.", s.ID, nsgID)
+			switch oc.Properties.NetworkProfile.PreconfiguredNSG {
+			case api.PreconfiguredNSGDisabled:
```

Comment:
nit: another `switch` with (basically) a single case that could be an `if/else`, and has potential to have no path executed, even though it "shouldn't ever happen" :p

Reply:
Done.